rfcs/002 - Custom Module Functions RFC.md

```diff
@@ -0,0 +1,80 @@
+
+YYYY-MM-DD
+
+#### Approvers
+
+- Luisfer De Pombo <luisfer@iasql.com>
+- Alejandro Guillen <alejandro@iasql.com>
+- Yolanda Robla <yolanda@iasql.com>
+
+### Implementation
+
+- [ ] Implemented: [One or more PRs](https://github.com/iasql/iasql-engine/some-pr-link-here) YYYY-MM-DD
+- [ ] Revoked/Superceded by: [RFC ###](./000 - RFC Template.md) YYYY-MM-DD
+
+## Author(s)
+
+- David Ellis <david@iasql.com>
+
+## Summary
+
+There are situations where the best (or only) way a module can provide a good experience for the service in question is via a custom Postgres function that is under-the-hood bound to behavior within the `iasql-engine`. Examples of this include querying DynamoDB/Redis/ElasticSearch/etc from within IaSQL, constructing a Lambda function payload zip from raw sources defined in SQL queries, executing an existing lambda function within a SQL query and operating on its resulting payload, etc.
+
+A mechanism to write custom functions in a generalized way across modules is proposed that will also simplify the logic within the [`scheduler.ts`](../src/services/scheduler.ts) file, along with a proposed Module interface change to shore up an issue that has recurred where custom functions get accidentally dropped from usage when module entity migration generation occurs.
+
+## Proposal
+
+The [`iasql_functions` module](../src/modules/0.0.14/iasql_functions) provides various `iasql_*` functions that under-the-hood are powered by a job processing system defined in [`scheduler.ts`](../src/services/scheduler.ts). This module and the scheduler are uncomfortably coupled via the [`enum` defined in the `iasql_functions` entity](../src/modules/0.0.14/iasql_functions/entity/index.ts) to link [these SQL functions](../src/modules/0.0.14/iasql_functions/sql/create_fns.sql) to [the actual functions](../src/services/iasql.ts) that do the work in question.
+
+This has always been a sore point architecturally as encapsulation of information is broken, and it is not truly possible to adjust the behavior of the functions in question without affecting prior module versions at the same time. It is also very difficult (though not impossible) to introduce new functions because 4 different files across the repository must be updated in sync to do so correctly, and can only be done in a way that would be considered a SemVer minor update, major is completely not allowed.
+
+It is proposed to deprecate/drop the `iasql_operation` entity and associated enum, and create a new graphile worker job type, called `rpc` and a new entity named `iasql_rpc` with no enum. Instead two fields would be added as an "addressing system" of module name and method name, which the module would implement within an `rpc` sub-object where all properties are the method names and the values are async functions that take a `Context` object as the first argument and then any arguments from the SQL side are added as follow-up parameters, all as strings similar to the current `params` array.
```

Comment:
Due to my background i have a very strong opinion about microservices. One approach could be dividing this single iasql engine inside smaller pieces, that have a very specific focus. The worker could be just one specific microservice, that is able to talk with others to requests specific actions. This framework adds better isolation and would be easier to develop and maintain, but would need some orchestration layer for communication (typically kubernetes)

Reply:
I also prefer microservices but hate Kubernetes. ;)

I don't think we're yet at the point where splitting things up like that makes sense, though.

1. There's still a tight coupling between the engine and the modules, with the interface between them still being in flux as we try to improve things there (as you yourself have seen the abstraction is not quite there yet for module development). Breaking that out will complicate iteration speed when we can't really afford that at the moment.
2. The "rpc" here is a remote procedure call in the classic sense, but it is between the Postgres database and the IaSQL engine as an RPC between two monoliths, essentially. The reason for this is that Postgres *is* our interface for end-users and the engine is an implementation detail that they don't need to pay much attention to, so as much of the API as possible is pushed into the database. This was chosen for a variety of reasons that should also become a technical blogpost ;) but the biggest one is that the fewer demands you place on the client to use your product, the more places they will be able to. You can theoretically do all of your cloud management with just a SQL client like ArcType this way, if you wanted to.
3. The operational simplicity of the engine as it currently stands is a huge bonus. Your own onboarding went smoother, I think, than if we had a hand-written runbook on how to set up a local microk8-or-whatever cluster and get the microservices running locally. (I dislike Kubernetes because it has tried to be all things to all people and is not opinionated enough to actually be any easier to manage than a classic nginx reverse proxy + app pool of hosts + db write main & read replicas that the vast majority of backend companies really ought to be using. If K8s came with a solid set of defaults and said "swap these out only if you know what you're doing" it would be a much better product.)

Microservices are great at 1) more efficiently allocating compute resources and 2) enforcing strict interface boundaries. We do not yet have an issue with (1) and we have not fully figured out what (2) should be, so I just think it's premature to move in that direction.

Reply:
Ok we have quite different point of view about Kubernetes :) But makes sense at least for the moment while there is no huge need to scale and distribute load